src/View/Form/DocumentContext.php

```diff
@@ -0,0 +1,365 @@
+            );
+        }
+        $this->_isCollection = (
+            is_array($entity) ||
+            $entity instanceof Traversable
+        );
+        $alias = $this->_rootName = $type->name();
+        $this->_types[$alias] = $type;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function primaryKey()
+    {
+        return ['id'];
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function isPrimaryKey($field)
+    {
+        $parts = explode('.', $field);
+        return array_pop($parts) === 'id';
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function isCreate()
+    {
+        $entity = $this->_context['entity'];
+        if (is_array($entity) || $entity instanceof Traversable) {
+            $entity = (new Collection($entity))->first();
+        }
+        if ($entity instanceof Document) {
+            return $entity->isNew() !== false;
+        }
+        return true;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function val($field)
+    {
+        $val = $this->_request->data($field);
+        if ($val !== null) {
+            return $val;
+        }
+        if (empty($this->_context['entity'])) {
+            return null;
+        }
+        $parts = explode('.', $field);
+        $entity = $this->entity($parts);
+        if ($entity instanceof Document) {
+            return $entity->get(array_pop($parts));
+        }
+    }
+
+    /**
+     * Get the entity that is closest to $path.
+     *
+     * @param array $path The to get an entity for.
+     * @return Cake\Datasource\EntityInterface|false The entity or false.
+     * @throws \RuntimeException when no entity can be found.
+     */
+    protected function entity($path)
+    {
+        if ($path === null) {
+            return $this->_context['entity'];
+        }
+
+        $oneElement = count($path) === 1;
+        if ($oneElement && $this->_isCollection) {
+            return false;
+        }
+        $entity = $this->_context['entity'];
+        if ($oneElement) {
+            return $entity;
+        }
+
+        if ($path[0] === $this->_rootName) {
+            $path = array_slice($path, 1);
+        }
+
+        $len = count($path);
+        $last = $len - 1;
+        for ($i = 0; $i < $len; $i++) {
+            $prop = $path[$i];
+            $next = $this->getProp($entity, $prop);
+            $isLast = ($i === $last);
+
+            if (!$isLast && $next === null && $prop !== '_ids') {
+                return false;
+            }
+
+            $isTraversable = (
+                is_array($next) ||
+                $next instanceof Traversable ||
+                $next instanceof Document
+            );
+            if ($isLast || !$isTraversable) {
+                return $entity;
+            }
+            $entity = $next;
+        }
+        throw new RuntimeException(sprintf(
+            'Unable to fetch property "%s"',
+            implode(".", $path)
+        ));
+    }
```

Comment:
I kind of like the duplication. It means we are more free to refactor each implementation without possibly breaking the other. Inheritance across repositories creeps me out sometimes.